### PR TITLE
MNT Additional behat tests

### PR DESCRIPTION
**Pull request by sabina-talipova**
## Description

Additional Behat tests to test user interaction with LinField using the keyboard.
Tests check the ability to create links and sort using "drag and drop".

## Manual testing steps

Not required

## Issues

- https://github.com/silverstripe/silverstripe-linkfield/issues/205

## Pull request checklist

- [x] The target branch is correct
    - See [picking the right version](https://docs.silverstripe.org/en/contributing/code/#picking-the-right-version)
- [x] All commits are relevant to the purpose of the PR (e.g. no debug statements, unrelated refactoring, or arbitrary linting)
    - Small amounts of additional linting are usually okay, but if it makes it hard to concentrate on the relevant changes, ask for the unrelated changes to be reverted, and submitted as a separate PR.
- [x] The commit messages follow our [commit message guidelines](https://docs.silverstripe.org/en/contributing/code/#commit-messages)
- [x] The PR follows our [contribution guidelines](https://docs.silverstripe.org/en/contributing/code/)
- [x] Code changes follow our [coding conventions](https://docs.silverstripe.org/en/contributing/coding_conventions/)
- [ ] This change is covered with tests (or tests aren't necessary for this change)
- [ ] Any relevant User Help/Developer documentation is updated; for impactful changes, information is added to the changelog for the intended release
- [x] CI is green

### CODE REVIEW

File: tests/behat/src/FeatureContext.php [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        $page = $this->getSession()->getPage();
+        $element = $page->find('css', $selector);
+        Assert::assertNotNull($element, sprintf('Element %s not found', $selector));
+        $element->focus();
+    }
+
+    /**
+     * @Given /^I type "([^"]+)" in the field$/
+     */
+    public function iTypeInField(string $data)
```
Comment (emteknetnz) [inline tests/behat/src/FeatureContext.php]:
```suggestion
    public function iTypeInTheField(string $data)
```
Method name should match Given regex

This method should be put into behat-extension BasicContext

File: tests/behat/src/FeatureContext.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        $driver = $this->getSession()->getDriver()->getWebDriver();
+        $driver->switchTo()->activeElement();
+        $keyboard = $driver->getKeyboard();
+        $keyboard->sendKeys([$data]);
+    }
+
+    /**
+     * @Given /^the active element is "([^"]+)"$/
+     */
+    public function isCurrentActiveElement(string $selector)
```
Comment (emteknetnz) [inline tests/behat/src/FeatureContext.php]:
```suggestion
    /**
     * @Given /^the active element should be "([^"]+)"$/
     */
    public function theActiveElementShouldBe(string $selector)
```
Updated both Given regex and the method name to be a little more standard

You'll need to update references in feature files too

This method should be put into behat-extension BasicContext

File: tests/behat/src/FeatureContext.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+use Facebook\WebDriver\WebDriverBy;
+use PHPUnit\Framework\Assert;
+use SilverStripe\BehatExtension\Context\SilverStripeContext;
+
+class FeatureContext extends SilverStripeContext
+{
+    /**
+     * @Given /^I focus on the "([^"]+)" element$/
+     */
+    public function theHasALinkInLinkField(string $selector)
```
Comment (emteknetnz) [inline tests/behat/src/FeatureContext.php]:
```suggestion
    public function iFocusOnTheElement(string $selector)
```

Method name should match Given regex

This method should be put into behat-extension BasicContext

File: tests/behat/src/FeatureContext.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
+<?php
+
+namespace SilverStripe\LinkField\Tests\Behat;
+
+use Facebook\WebDriver\WebDriverBy;
+use PHPUnit\Framework\Assert;
+use SilverStripe\BehatExtension\Context\SilverStripeContext;
+
+class FeatureContext extends SilverStripeContext
```
Comment (emteknetnz) [inline tests/behat/src/FeatureContext.php]: Move the methods in this class to behat-extension BasicContext and delete this file

File: behat.yml [hunk 5]
```diff
@@ -10,13 +10,13 @@ default:
       paths:
         - '%paths.modules.linkfield%/tests/behat/features'
       contexts:
-        - SilverStripe\Framework\Tests\Behaviour\FeatureContext
```
Comment (emteknetnz) [inline behat.yml]: Just leave this in

File: behat.yml [hunk 6]
```diff
@@ -10,13 +10,13 @@ default:
...
         - SilverStripe\Framework\Tests\Behaviour\CmsFormsContext
         - SilverStripe\Framework\Tests\Behaviour\CmsUiContext
         - SilverStripe\BehatExtension\Context\BasicContext
         - SilverStripe\BehatExtension\Context\LoginContext
+        - SilverStripe\LinkField\Tests\Behat\FeatureContext
```
Comment (emteknetnz) [inline behat.yml]:
```suggestion
```

File: behat.yml [hunk 7]
```diff
@@ -10,13 +10,13 @@ default:
...
         - SilverStripe\BehatExtension\Context\FixtureContext:
```
Comment (emteknetnz) [inline behat.yml]: Could you change the FixtureContext bit to have the **exact** same indentation as [this](https://github.com/silverstripe/silverstripe-admin/blob/2/behat.yml#L21) - it's very weird and it's like that for some reason that I forget, though I do know it definately works

File: tests/behat/features/accessability-linkfield.feature [hunk 8]
```diff
@@ -0,0 +1,128 @@
+@retry
+Feature: Accessibility Tests
+As a content editor
+I want to create and interact with LinkField using the keyboard
+
+  Background:
+    Given I add an extension "SilverStripe\FrameworkTest\LinkField\Extensions\LinkPageExtension" to the "Page" class
+      And I go to "/dev/build?flush"
```
Comment (emteknetnz) [inline tests/behat/features/accessability-linkfield.feature]:
```suggestion
    Given I add an extension "SilverStripe\FrameworkTest\LinkField\Extensions\LinkPageExtension" to the "Page" class
    And I go to "/dev/build?flush"
```
Use the same indentation for all the statements within Background/Scenario - that's generally what's done in other feature files in the code base

The this for the entire file

File: tests/behat/features/accessability-linkfield.feature [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+      Then I should see "External URL is required" in the ".modal-content" element
+      And I press the "Tab" key globally
+      And I press the "Tab" key globally
+      And I press the "Tab" key globally
+      And I press the "Tab" key globally
+      And the active element is ".close"
+      And I press the "Enter" key globally
+      And I wait for 2 seconds
+
+      # Test accessability of the LinkField drag and drop
```
Comment (emteknetnz) [inline tests/behat/features/accessability-linkfield.feature]:
```suggestion
      # Test accessibility of the LinkField keyboard sorting
```
Drag and drop is using the mouse :-)

File: tests/behat/features/accessability-linkfield.feature [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+      # Create SiteTreeLink in LinkField using keyboard
+
+      Then I focus on the "[data-field-id='Form_EditForm_HasOneLink'] button" element
+      And I press the "Enter" key globally
+      And I press the "Enter" key globally
+      And I wait for 2 seconds
+      And I should see "Page on this site" in the ".modal-header" element
+      And I wait for 2 seconds
+
+      # Test accessability of the modal form
```
Comment (emteknetnz) [inline tests/behat/features/accessability-linkfield.feature]:
```suggestion
      # Test accessibility of the modal form
```

**Review by emteknetnz — CHANGES_REQUESTED** [review]

File: tests/behat/src/FeatureContext.php (hunk 1 above)
Comment (sabina-talipova) [reply tests/behat/src/FeatureContext.php]: Updated. Moved in `behat-extension`. [See](https://github.com/silverstripe/silverstripe-behat-extension/pull/263/files)

File: tests/behat/src/FeatureContext.php (hunk 2 above)
Comment (sabina-talipova) [reply tests/behat/src/FeatureContext.php]: Updated. Moved in `behat-extension`. [See](https://github.com/silverstripe/silverstripe-behat-extension/pull/263/files)

File: tests/behat/src/FeatureContext.php (hunk 3 above)
Comment (sabina-talipova) [reply tests/behat/src/FeatureContext.php]: Updated. Moved in `behat-extension`. [See](https://github.com/silverstripe/silverstripe-behat-extension/pull/263/files)

File: tests/behat/src/FeatureContext.php (hunk 4 above)
Comment (sabina-talipova) [reply tests/behat/src/FeatureContext.php]: Updated.

File: behat.yml (hunk 5 above)
Comment (sabina-talipova) [reply behat.yml]: Updated.

File: behat.yml (hunk 6 above)
Comment (sabina-talipova) [reply behat.yml]: Updated

File: behat.yml (hunk 7 above)
Comment (sabina-talipova) [reply behat.yml]: Updated.

File: tests/behat/features/accessability-linkfield.feature (hunk 8 above)
Comment (sabina-talipova) [reply tests/behat/features/accessability-linkfield.feature]: Updated.

File: tests/behat/features/accessability-linkfield.feature (hunk 9 above)
Comment (sabina-talipova) [reply tests/behat/features/accessability-linkfield.feature]: Updated

File: tests/behat/features/accessability-linkfield.feature (hunk 10 above)
Comment (sabina-talipova) [reply tests/behat/features/accessability-linkfield.feature]: Updated.

**Comment by sabina-talipova** [comment]
Test passed: https://github.com/sabina-talipova/recipe-kitchen-sink/actions/runs/7924765855/job/21636895068

**Review by emteknetnz — APPROVED** [review]
Approved, will merge when https://github.com/silverstripe/silverstripe-behat-extension/pull/263 is merged and tagged

**Comment by emteknetnz** [comment]
Have merged and released behat-extenstion pr, re-running behat tests on this pr now

**Pull request merged**